File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/operation/meta/schema/resolver/ResolveArgs.java
package ru.citeck.ecos.records3.record.operation.meta.schema.resolver;

import lombok.Data;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import ru.citeck.ecos.commons.utils.MandatoryParam;
import ru.citeck.ecos.records3.RecordRef;
import ru.citeck.ecos.records3.record.operation.meta.schema.AttSchema;
import ru.citeck.ecos.records3.record.operation.meta.schema.SchemaRootAtt;
import ru.citeck.ecos.records3.source.common.AttMixin;

import java.util.*;

@Data
@RequiredArgsConstructor
public class ResolveArgs {

    @NotNull
    private final List<Object> values;
    @NotNull
    private final String sourceId;
    @NotNull
    private final List<RecordRef> valueRefs;
    @NotNull
    private final AttSchema schema;
    @NotNull
    private final List<AttMixin> mixins;

    private final boolean rawAtts;

    public static Builder create() {
        return new Builder();
    }

    public static class Builder {

        private String sourceId = "";
        private List<Object> values = new ArrayList<>();
        private List<RecordRef> valueRefs = Collections.emptyList();
        private AttSchema schema;

        @Getter
        private boolean rawAtts;

        private List<AttMixin> mixins = new ArrayList<>();

        public Builder setSourceId(String sourceId) {
            this.sourceId = sourceId;
            return this;
        }

        public Builder setRawAtts(boolean value) {
            rawAtts = value;
            return this;
        }

        @NotNull
        public Builder setValues(@Nullable List<Object> values) {
            this.values = values != null ? new ArrayList<>(values) : Collections.emptyList();
            return this;
        }

        @NotNull
        public Builder setValueRefs(@Nullable List<RecordRef> valueRefs) {
            this.valueRefs = valueRefs != null ? new ArrayList<>(valueRefs) : Collections.emptyList();
            return this;
        }

        @NotNull
        public Builder setMixins(@Nullable Collection<AttMixin> mixins) {
            if (mixins == null) {
                this.mixins = new ArrayList<>();
            } else {
                this.mixins = new ArrayList<>(mixins);
            }
            return this;
        }

        @NotNull
        public Builder setAtts(@Nullable List<SchemaRootAtt> attributes) {
            this.schema = new AttSchema(attributes != null ? new ArrayList<>(attributes) : Collections.emptyList());
            return this;
        }

        @NotNull
        public Builder setAtt(@Nullable SchemaRootAtt attribute) {
            return setAtts(attribute != null ? Collections.singletonList(attribute) : null);
        }

        public Builder setSchema(AttSchema schema) {
            this.schema = schema;
            return this;
        }

        public ResolveArgs build() {

            MandatoryParam.check("schema", schema);
            MandatoryParam.check("values", values);

            if (!valueRefs.isEmpty() && valueRefs.size() != values.size()) {
                throw new RuntimeException("valueRefs should have same size with values");
            }
            return new ResolveArgs(values, sourceId, valueRefs, schema, mixins, rawAtts);
        }
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/atts/schema/resolver/ResolveArgs.kt
package ru.citeck.ecos.records3.record.atts.schema.resolver

import ru.citeck.ecos.records3.record.atts.schema.SchemaAtt
import ru.citeck.ecos.records3.record.mixin.EmptyMixinContext
import ru.citeck.ecos.records3.record.mixin.MixinContext
import ru.citeck.ecos.webapp.api.entity.EntityRef

data class ResolveArgs(
    val values: List<Any?>,
    val sourceId: String,
    val defaultValueRefs: List<EntityRef>,
    val attributes: List<SchemaAtt>,
    val mixinCtx: MixinContext,
    val rawAtts: Boolean
) {

    companion object {

        @JvmStatic
        fun create(): Builder {
            return Builder()
        }

        @JvmStatic
        fun create(builder: Builder.() -> Unit): ResolveArgs {
            val builderObj = Builder()
            builder.invoke(builderObj)
            return builderObj.build()
        }
    }

    fun copy(): Builder {
        return Builder(this)
    }

    fun copy(builder: Builder.() -> Unit): ResolveArgs {
        val builderObj = Builder(this)
        builder.invoke(builderObj)
        return builderObj.build()
    }

    class Builder() {

        var values: List<Any?> = emptyList()
        var sourceId: String = ""
        var defaultValueRefs: List<EntityRef> = emptyList()
        var attributes: List<SchemaAtt> = emptyList()
        var mixinCtx: MixinContext = EmptyMixinContext
        var rawAtts = false

        constructor(base: ResolveArgs) : this() {
            values = ArrayList(base.values)
            sourceId = base.sourceId
            defaultValueRefs = ArrayList(base.defaultValueRefs)
            attributes = ArrayList(base.attributes)
            mixinCtx = base.mixinCtx
            rawAtts = base.rawAtts
        }

        fun withValues(values: List<*>): Builder {
            this.values = values
            return this
        }

        fun withSourceId(sourceId: String): Builder {
            this.sourceId = sourceId
            return this
        }

        fun withDefaultValueRefs(defaultValueRefs: List<EntityRef>): Builder {
            this.defaultValueRefs = defaultValueRefs
            return this
        }

        fun withAttribute(attribute: SchemaAtt): Builder {
            return withAttributes(listOf(attribute))
        }

        fun withAttributes(attributes: List<SchemaAtt>): Builder {
            this.attributes = attributes
            return this
        }

        fun withMixinContext(mixinCtx: MixinContext): Builder {
            this.mixinCtx = mixinCtx
            return this
        }

        fun withRawAtts(rawAtts: Boolean): Builder {
            this.rawAtts = rawAtts
            return this
        }

        fun build(): ResolveArgs {
            if (defaultValueRefs.isNotEmpty() && defaultValueRefs.size != values.size) {
                throw RuntimeException("defaultValueRefs should have same size with values")
            }
            return ResolveArgs(
                values = values,
                sourceId = sourceId,
                defaultValueRefs = defaultValueRefs,
                attributes = attributes,
                mixinCtx = mixinCtx,
                rawAtts = rawAtts
            )
        }
    }
}
